src/store/index-level.ts

```diff
@@ -1,213 +1,435 @@
+   * @param indexes - (key-value pairs) to be included as part of indexing this item. Must include at least one indexing property.
+   * @param options IndexLevelOptions that include an AbortSignal.
    */
   async put(
     tenant: string,
-    dataId: string,
-    indexes: { [property: string]: unknown },
+    itemId: string,
+    indexes: KeyValues,
     options?: IndexLevelOptions
   ): Promise<void> {
-    const partition = await executeUnlessAborted(this.db.partition(tenant), options?.signal);
-    indexes = flatten(indexes);
-
-    const operations: LevelWrapperBatchOperation<string>[] = [ ];
-
-    // create an index entry for each property in the `indexes`
-    for (const propertyName in indexes) {
-      const propertyValue = indexes[propertyName];
-
-      // NOTE: appending data ID after (property + value) serves two purposes:
-      // 1. creates a unique entry of the property-value pair per data/object
-      // 2. when we need to delete all indexes of a given data ID (`delete()`), we can reconstruct the index keys and remove the indexes efficiently
-      //
-      // example keys (\u0000 is just shown for illustration purpose because it is the delimiter used to join the string segments below):
-      // 'interface\u0000"Records"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      // 'method\u0000"Write"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      // 'schema\u0000"http://ud4kyzon6ugxn64boz7v"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      // 'dataCid\u0000"bafkreic3ie3cxsblp46vn3ofumdnwiqqk4d5ah7uqgpcn6xps4skfvagze"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      // 'dateCreated\u0000"2023-05-25T18:23:29.425008Z"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
-      const key = this.join(propertyName, this.encodeValue(propertyValue), dataId);
-      operations.push({ type: 'put', key, value: dataId });
+
+    // ensure we have something valid to index, remove any empty strings from index properties
+    removeEmptyStrings(indexes);
+    if (isEmptyObject(indexes)) {
+      throw new DwnError(DwnErrorCode.IndexMissingIndexableProperty, 'Index must include at least one valid indexable property');
+    }
+
+    const indexOps: LevelWrapperBatchOperation<string>[] = [];
+
+    // create an index entry for each property index
+    // these indexes are all sortable lexicographically.
+    for (const indexName in indexes) {
+      const indexValue = indexes[indexName];
+      // the key is indexValue followed by the itemId as a tie-breaker.
+      // for example if the property is messageTimestamp the key would look like:
+      // '"2023-05-25T18:23:29.425008Z"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
+      const key = IndexLevel.keySegmentJoin(IndexLevel.encodeValue(indexValue), itemId);
+      const item: IndexedItem = { itemId, indexes };
```

Comment:
Guess I will find out as I read on why indexes need to be included for every index write/push. Immediate reaction assuming I am reading it correctly is that this seems rather heavy. e.g. for every RecordsWrite, we could be looking at close to 20 key-value pairs stored repetitively ~20 times.

Reply:
Discussed offline, would be nice to confirm it being too costly to look up the indexes for every messageCid, if it's not "too" costly, it's probably worth just doing the lookup and avoid storing the same indexes again and again.

But we can definitely investigate in a separate PR if it's not low hanging fruit.

Reply:
Separate PR it is.

Reply:
Filed a bug https://github.com/TBD54566975/dwn-sdk-js/issues/631 to track this.